hw/drivers/trng/trng_k64f/src/trng_k64f.c

```diff
@@ -0,0 +1,175 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <string.h>
+#include "mcu/cmsis_nvic.h"
+#include "fsl_rnga.h"
+#include "trng/trng.h"
+#include "trng_k64f/trng_k64f.h"
+
+static uint8_t rng_cache[ MYNEWT_VAL(K64F_TRNG_CACHE_LEN) ];
+static uint16_t rng_cache_out;
+static uint16_t rng_cache_in;
+static os_stack_t *pstack;
+static bool running;
+
+#define RNGA_POLLER_PRIO (8)
+#define RNGA_POLLER_STACK_SIZE OS_STACK_ALIGN(64)
+static struct os_task poller_task;
+
+static inline void
+k64f_rnga_start(void)
+{
+    RNGA_SetMode(RNG, kRNGA_ModeNormal);
+    running = true;
+}
+
+static inline void
+k64f_rnga_stop(void)
+{
+   RNGA_SetMode(RNG, kRNGA_ModeSleep);
+   running = false;
+}
+
+static size_t
+k64f_trng_read(struct trng_dev *trng, void *ptr, size_t size)
+{
+    os_sr_t sr;
+    size_t num_read;
+
+    OS_ENTER_CRITICAL(sr);
+
+    if (rng_cache_out <= rng_cache_in) {
+        size = min(size, rng_cache_in - rng_cache_out);
+        memcpy(ptr, &rng_cache[rng_cache_out], size);
+        num_read = size;
+    } else if (rng_cache_out + size <= sizeof(rng_cache)) {
+        memcpy(ptr, &rng_cache[rng_cache_out], size);
+        num_read = size;
+    } else {
+        num_read = sizeof(rng_cache) - rng_cache_out;
+        memcpy(ptr, &rng_cache[rng_cache_out], num_read);
+
+        size -= num_read;
+        ptr += num_read;
+
+        size = min(size, rng_cache_in);
+        memcpy(ptr, rng_cache, size);
+        num_read += size;
+    }
+
+    rng_cache_out = (rng_cache_out + num_read) % sizeof(rng_cache);
+
+    if (num_read > 0) {
+        k64f_rnga_start();
+    }
+
+    OS_EXIT_CRITICAL(sr);
+
+    return num_read;
+}
+
+static uint32_t
+k64f_trng_get_u32(struct trng_dev *trng)
+{
+    union {
+        uint32_t v32;
+        uint8_t v8[4];
+    } val;
+    size_t num;
+
+    num = k64f_trng_read(trng, &val.v8, sizeof(val.v8));
+    while (num < sizeof(val.v8)) {
+        os_sched(NULL);
+        num += k64f_trng_read(trng, &val.v8[num], sizeof(val.v8) - num);
+    }
+
+    return val.v32;
+}
+
+static void
+rnga_poller_handler(void *arg)
+{
+    int8_t i;
+    uint8_t data[4];
+    int rc;
+
+    while (1) {
+        if (running) {
+            rc = RNGA_GetRandomData(RNG, data, sizeof(uint32_t));
+            if (rc == 0) {
+                for (i = 0; i < 4; i++) {
+                    rng_cache[rng_cache_in++] = data[i];
```

Comment:
there is possible time between rng_cache_in get incremented and checked against buffer size. During that time value rng_cache_in is invalid (out of range). If during this time k64f_trng_read() reads it in line 58 it will copy last byte that was not just generated.

Reply:
This is a good one! `k64f_trng_read` was disabling interrupts because it was based on nrf52/stm32 which generate ints when there is RNG data (and are also affected by this bug). I removed the interrupt disabling code and added a simple mutex based access to `rng_cache_*`.